### PR TITLE
Morph - allow for full page morphs

### PR DESCRIPTION
**Pull request by VicGUTT**
Totally forgot this Saturday would be Christmas but [**my word is my bond**](https://github.com/alpinejs/alpine/discussions/2511#discussioncomment-1849933).

This PR should allow for full HTML page morphs as discussed in https://github.com/alpinejs/alpine/discussions/2511.

The failing morph tests should be fixed in the following PR : https://github.com/alpinejs/alpine/pull/2528.

Happy holidays :v:

### CODE REVIEW

File: packages/morph/src/morph.js [hunk 1]
```diff
@@ -61,8 +61,30 @@ function assignOptions(options = {}) {
     debug = options.debug || false
 }
 
-function createElement(html) {
-    return document.createRange().createContextualFragment(html).firstElementChild
+function createElement(to, from = null) {
+    const DOCUMENT_FRAGMENT_NODE = 11
+
+    if (from && typeof to === 'string' && ['#document', 'HTML', 'BODY'].includes(from.nodeName)) {
```
[the remark this answers is not in the document]
Comment (VicGUTT) [reply packages/morph/src/morph.js:67]: Hum... I'm not sure I fully understand the question but I think the answer is : yes ! ^^

On the quoted line, if the _from_ element being morphed into the _to_ element is a Document object, the HTML tag or the Body tag, then it ensures the _to_ element also has the same tag hierarchy.

That piece of code was [mostly taken from Morphdom as is](https://github.com/patrick-steele-idem/morphdom/blob/84f28d0438f432cdef8cd869cd67ab916c029f2e/src/morphdom.js#L26,L36) and from what I understand it prevents DOM elements from going from `<html><head></head><body><div>Hey</div></body></html>` to simply `<div>Hey</div>` whereas the opposite is fine.

Examples:

```js
createElement('<div>Hey</div>'); // <div>Hey</div>
createElement('<div>Hey</div>', document.createElement('span')); // <div>Hey</div>
createElement('<div>Hey</div>', document); // <html><head></head><body><div>Hey</div></body></html>

createElement('<html><head></head><body><div>Hey</div></body></html>'); // <div>Hey</div>
createElement('<html><head></head><body><div>Hey</div></body></html>', document.createElement('span')); // <div>Hey</div>
createElement('<html><head></head><body><div>Hey</div></body></html>', document); // <html><head></head><body><div>Hey</div></body></html>
```

**Comment by rafales** [comment]
Is there any reason we are not allowing a call to `Alpine.morph()` with an actual node but instead allow only html? I'm looking to call morph on body. I'm using DOMParser to parse fetch response. With the current state of thins I still have to serialize `body` back to HTML, which then again will be deserialized into a node. Kinda wasteful.

**Comment by cleberapisdesign** [comment]
May I contribute with an use-case?

I'm currently building a CMS-like interface where the user edits content on a richtext editor and sees it fully rendered (with custom scripts and styles) real-time inside of an iframe. 

Firstly I was injecting my editor's output into iframe's `body.innerHTML` with some `postMessage` trickery. It was all good until I've added some features that need to touch `<head>` element (to add or edit style and scripts dynamically based on input content). Changing <head>s innerHTML makes the browser redownload all linked stylesheets and relayout the whole page (even if the incoming head code is the same as the one that is applied, which is a non-trivial comparision to make).

Now I'm trying to use Alpine morph plugin to change only what's actually changed inside head and body tags, (or inside the whole html, whichever is easier/possible). The problem is, the first time I run alpine morph on one of these tags, the tag gets replaced with its content. e. g. `<body><div>hello</div></body>` becomes `<div>hello</div>`, so subsequent calls don't find body element and fail. Will try and reinsert body tag on each morph operation, but I think that's not ideal.

**Comment by joshhanley** [comment]
Thanks for the PR! We're going to close this as morph has had a big restructure since this PR.

**Pull request closed without merging**